### PR TITLE
Add newer syntax to detached HEAD check

### LINKED ISSUE #38876: git module update fails if in detached HEAD state
##### ISSUE TYPE
 - Bug Report

##### COMPONENT NAME

git module

##### ANSIBLE VERSION

```
# ansible --version
ansible 2.5.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/root/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python2.7/site-packages/ansible
  executable location = /usr/bin/ansible
  python version = 2.7.5 (default, Aug  4 2017, 00:39:18) [GCC 4.8.5 20150623 (Red Hat 4.8.5-16)]
```

##### CONFIGURATION

no output of `ansible-config dump --only-changed`

##### OS / ENVIRONMENT

Managing from [CentOS container](https://github.com/ansible/awx/blob/devel/tools/docker-compose/Dockerfile)

```
# cat /proc/version
Linux version 4.9.87-linuxkit-aufs (root@95fa5ec30613) (gcc version 6.4.0 (Alpine 6.4.0) ) #1 SMP Wed Mar 14 15:12:16 UTC 2018
```

Doing this as all local actions

##### SUMMARY
If I try to update when starting out in a detached head state, then the git module fails on a command that syntactically looks like a bug.

##### STEPS TO REPRODUCE

The playbook command:

```
# pwd
/projects/_845__callback_testing_playbooks
# git status
HEAD detached at 1ebf2f9
nothing to commit, working tree clean
# ANSIBLE_STDOUT_CALLBACK=json ansible-playbook -i "localhost," /awx_devel/pu_test.yml
```

```yaml
---
- hosts: all
  connection: local
  gather_facts: false
  tasks:

      - name: update project using git
        git:
          dest: "/projects/_845__callback_testing_playbooks"
          repo: "https://github.com/AlanCoding/callback-testing-playbooks.git"
          version: "HEAD"
          force: "false"
        register: scm_result
```

You can find commit history here

https://github.com/AlanCoding/callback-testing-playbooks

This source tree is in a detached head state on an old commit

```
# git remote -v
origin	https://github.com/AlanCoding/callback-testing-playbooks.git (fetch)
origin	https://github.com/AlanCoding/callback-testing-playbooks.git (push)
```

I should note that I used the git module to get the source tree in this state to begin with. I can dig to produce more details about how to get this stating state if needed.

##### EXPECTED RESULTS
Updates to HEAD of `origin/master`

I have a reference folder that is on the `master` branch. The output looks like this:

```json
{
    "plays": [
        {
            "play": {
                "id": "0242ac12-0006-ba05-8c27-000000000006", 
                "name": "all"
            }, 
            "tasks": [
                {
                    "hosts": {
                        "localhost": {
                            "_ansible_no_log": false, 
                            "_ansible_parsed": true, 
                            "after": "c1b69a96653cd16a130dd5cb264a6bee8a91662a", 
                            "before": "c1b69a96653cd16a130dd5cb264a6bee8a91662a", 
                            "changed": false, 
                            "invocation": {
                                "module_args": {
                                    "accept_hostkey": false, 
                                    "archive": null, 
                                    "bare": false, 
                                    "clone": true, 
                                    "depth": null, 
                                    "dest": "/projects/_844__callback_testing_playbooks", 
                                    "executable": null, 
                                    "force": false, 
                                    "key_file": null, 
                                    "recursive": true, 
                                    "reference": null, 
                                    "refspec": null, 
                                    "remote": "origin", 
                                    "repo": "https://github.com/AlanCoding/callback-testing-playbooks", 
                                    "ssh_opts": null, 
                                    "track_submodules": false, 
                                    "umask": null, 
                                    "update": true, 
                                    "verify_commit": false, 
                                    "version": "HEAD"
                                }
                            }, 
                            "remote_url_changed": false
                        }
                    }, 
                    "task": {
                        "id": "0242ac12-0006-ba05-8c27-000000000008", 
                        "name": "update project using git"
                    }
                }
            ]
        }
    ], 
    "stats": {
        "localhost": {
            "changed": 0, 
            "failures": 0, 
            "ok": 1, 
            "skipped": 0, 
            "unreachable": 0
        }
    }
}
```

##### ACTUAL RESULTS

```json
{
    "plays": [
        {
            "play": {
                "id": "0242ac12-0006-1cd2-ba28-000000000006", 
                "name": "all"
            }, 
            "tasks": [
                {
                    "hosts": {
                        "localhost": {
                            "_ansible_no_log": false, 
                            "_ansible_parsed": true, 
                            "changed": false, 
                            "cmd": "/usr/bin/git reset --hard origin/1ebf2f923de524d836503be96aff1b7202afba97 --", 
                            "invocation": {
                                "module_args": {
                                    "accept_hostkey": false, 
                                    "archive": null, 
                                    "bare": false, 
                                    "clone": true, 
                                    "depth": null, 
                                    "dest": "/projects/_845__callback_testing_playbooks", 
                                    "executable": null, 
                                    "force": false, 
                                    "key_file": null, 
                                    "recursive": true, 
                                    "reference": null, 
                                    "refspec": null, 
                                    "remote": "origin", 
                                    "repo": "https://github.com/AlanCoding/callback-testing-playbooks.git", 
                                    "ssh_opts": null, 
                                    "track_submodules": false, 
                                    "umask": null, 
                                    "update": true, 
                                    "verify_commit": false, 
                                    "version": "HEAD"
                                }
                            }, 
                            "msg": "Failed to checkout branch 1ebf2f923de524d836503be96aff1b7202afba97", 
                            "rc": 128, 
                            "stderr": "fatal: Failed to resolve 'origin/1ebf2f923de524d836503be96aff1b7202afba97' as a valid revision.\n", 
                            "stderr_lines": [
                                "fatal: Failed to resolve 'origin/1ebf2f923de524d836503be96aff1b7202afba97' as a valid revision."
                            ], 
                            "stdout": "", 
                            "stdout_lines": []
                        }
                    }, 
                    "task": {
                        "id": "0242ac12-0006-1cd2-ba28-000000000008", 
                        "name": "update project using git"
                    }
                }
            ]
        }
    ], 
    "stats": {
        "localhost": {
            "changed": 0, 
            "failures": 1, 
            "ok": 0, 
            "skipped": 0, 
            "unreachable": 0
        }
    }
}
```

Looks like that command is probably coming from

https://github.com/ansible/ansible/blob/cdd21e2170b9a47d84624c8c2f792a66f2b0fe57/lib/ansible/modules/source_control/git.py#L819

It certainly doesn't run

```
# /usr/bin/git reset --hard origin/1ebf2f923de524d836503be96aff1b7202afba97 --
fatal: Failed to resolve 'origin/1ebf2f923de524d836503be96aff1b7202afba97' as a valid revision.
```

Anyway, this doesn't appear to match with the intent of my playbook either. I wanted to checkout the latest, and if it was going to fetch the hash for the latest master HEAD from origin, that would be `c1b69a96653cd16a130dd5cb264a6bee8a91662a`. The hash it does try to reset to is the current local revision, which is out-of-date.

----

more details, which source code suggests should be relevant.

```
# cat .git/refs/remotes/origin/HEAD
ref: refs/remotes/origin/master
```

```
# cat .git/HEAD
1ebf2f923de524d836503be96aff1b7202afba97
```

```
# git branch --no-color -a
* (HEAD detached at 1ebf2f9)
  master
  remotes/origin/HEAD -> origin/master
  remotes/origin/master
```

### PR DESCRIPTION
**Pull request by AlanCoding**
Fixes https://github.com/ansible/ansible/issues/38876

### CODE REVIEW

**Comment by AlanCoding** [comment]
I tested in my environment in that issue, and got a success, now I'm at master.

```json
{
    "custom_stats": {}, 
    "plays": [
        {
            "play": {
                "id": "0242ac12-0006-6be7-21b5-000000000006", 
                "name": "all"
            }, 
            "tasks": [
                {
                    "hosts": {
                        "localhost": {
                            "_ansible_no_log": false, 
                            "_ansible_parsed": true, 
                            "action": "git", 
                            "after": "c1b69a96653cd16a130dd5cb264a6bee8a91662a", 
                            "before": "1ebf2f923de524d836503be96aff1b7202afba97", 
                            "changed": true, 
                            "invocation": {
                                "module_args": {
                                    "accept_hostkey": false, 
                                    "archive": null, 
                                    "bare": false, 
                                    "clone": true, 
                                    "depth": null, 
                                    "dest": "/projects/_845__callback_testing_playbooks", 
                                    "executable": null, 
                                    "force": false, 
                                    "key_file": null, 
                                    "recursive": true, 
                                    "reference": null, 
                                    "refspec": null, 
                                    "remote": "origin", 
                                    "repo": "https://github.com/AlanCoding/callback-testing-playbooks.git", 
                                    "ssh_opts": null, 
                                    "track_submodules": false, 
                                    "umask": null, 
                                    "update": true, 
                                    "verify_commit": false, 
                                    "version": "HEAD"
                                }
                            }, 
                            "remote_url_changed": false
                        }
                    }, 
                    "task": {
                        "id": "0242ac12-0006-6be7-21b5-000000000008", 
                        "name": "update project using git"
                    }
                }
            ]
        }
    ], 
    "stats": {
        "localhost": {
            "changed": 1, 
            "failures": 0, 
            "ok": 1, 
            "skipped": 0, 
            "unreachable": 0
        }
    }
}
```

Tested by

```
pip uninstall ansible
pip install git+https://github.com/AlanCoding/ansible.git@detached_at
```

File: lib/ansible/modules/source_control/git.py [hunk 1]
```diff
@@ -590,7 +590,7 @@ def is_local_branch(git_path, module, dest, branch):
 def is_not_a_branch(git_path, module, dest):
     branches = get_branches(git_path, module, dest)
     for branch in branches:
-        if branch.startswith('* ') and ('no branch' in branch or 'detached from' in branch):
+        if branch.startswith('* ') and ('no branch' in branch or 'detached from' or 'detached at' in branch):
```
Comment (mkrizek) [inline lib/ansible/modules/source_control/git.py]: I think it should be:
```
if branch.startswith('* ') and ('no branch' in branch or 'detached from' in branch or 'detached at' in branch)
```

Comment (AlanCoding) [reply lib/ansible/modules/source_control/git.py]: d'oh, thanks!

**Review by mkrizek — APPROVED** [review]

**Review by Akasurde — APPROVED** [review]
LGTM

**Pull request merged**